### PR TITLE
Soen17 admin roles

### PR DESCRIPTION
**Pull request by SergioAAI**
# Description

Implements [SOEN-17](https://soen390team8.atlassian.net/browse/SOEN-17)

- Added page displaying all 5 roles with the users associated to it
- Added modal to change a user's role
- Page is only visible if logged in as an Admin

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Only Manual Test

User changes in the front-end are reflected in the data

Page shows access denied if not logged in as an admin

# Screenshots

image

### CODE REVIEW

File: backend/controllers/userController.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+const db = require('../config/database');
+
+const getAllRoles = async (req, res, next) => {
+  const roles = await db.query(
+    'SELECT AccountId, FirstName, LastName, Role FROM MinicapDatabase.dbo.Users t',
+    {
+      logging: console.log,
+      type: QueryTypes.SELECT,
+    }
+  );
```
Comment (jfv-2000) [inline backend/controllers/userController.js]: Use the sequelize format to fetch the selected users just to keep the coding convention constant

File: backend/controllers/userController.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
+const { QueryTypes } = require('sequelize');
+const db = require('../config/database');
+
+const getAllRoles = async (req, res, next) => {
```
Comment (jfv-2000) [inline backend/controllers/userController.js]: I found the name to be a bit confusing, is it possible to add a comment on this, or to change the name of the constant, (e.g: getUsersOfAllRoles) ?

File: backend/controllers/adminController.js [hunk 3]
```diff
@@ -1,21 +1,32 @@
...
+const User = require('../models/user');
 
 //Update role of a user
 function updateRole(req, res) {
-    User.update({
-        Role: req.body.role
-    }, {
+  console.log(req.body);
+
+  if (req.body.NewRole == req.body.OldRole) {
```
Comment (jfv-2000) [inline backend/controllers/adminController.js]: We should keep the styling convention constants for the attributes of the "req.body". In this case, req.body.NewRole should be newRole, req.body.OldRole should be oldRole. The name would have to be adjusted in the front-end in this case also.

File: backend/controllers/adminController.js [hunk 4]
```diff
@@ -1,21 +1,32 @@
...
+    res.status(200);
+  } else {
+    User.update(
+      {
+        Role: req.body.NewRole,
```
Comment (jfv-2000) [inline backend/controllers/adminController.js]: Same as comment mentioning styling convention

File: backend/controllers/adminController.js [hunk 5]
```diff
@@ -1,21 +1,32 @@
...
         where: {
-            Email: req.body.email
-        }
-    }).then(result => {
-        console.log("Role successfully updated !");
-        res.status(200).send("Role successfully updated !");
-    }).catch(err => {
-        res.status(500).send("Error:"+err);
-    })
+          AccountId: req.body.AccountId,
```
Comment (jfv-2000) [inline backend/controllers/adminController.js]: Same as comment mentioning styling convention

File: backend/package.json [hunk 6]
```diff
@@ -18,7 +18,8 @@
     "jade": "^1.11.0",
     "jsonwebtoken": "^8.5.1",
     "mssql": "^7.3.0",
-    "sequelize": "^5.22.3"
+    "sequelize": "^5.22.3",
+    "tedious": "^14.1.0"
```
Comment (jfv-2000) [inline backend/package.json]: Remove Tedious if not utilized

File: backend/controllers/adminController.js [hunk 7]
```diff
@@ -1,21 +1,32 @@
-const User = require("../models/user");
+const User = require('../models/user');
 
 //Update role of a user
 function updateRole(req, res) {
-    User.update({
-        Role: req.body.role
-    }, {
+  console.log(req.body);
```
Comment (jfv-2000) [inline backend/controllers/adminController.js]: Remove unnecessary console logs

File: backend/controllers/adminController.js (hunk 5 above)
Comment (jfv-2000) [reply backend/controllers/adminController.js]: Tested the update function locally, currently not working properly due to line 16. The accountId of the user should be used to verify if the user has the rights to update roles as an admin. The email should be used in this case to properly update the role of a user.

File: backend/controllers/adminController.js (hunk 7 above)
Comment (NicholasYip) [inline backend/controllers/adminController.js]: you might have forgotten to remove this after debugging

File: backend/controllers/adminController.js (hunk 3 above)
Comment (NicholasYip) [inline backend/controllers/adminController.js]: should be a `===`

File: backend/controllers/userController.js [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+  const roles = await db.query(
+    'SELECT AccountId, FirstName, LastName, Role FROM MinicapDatabase.dbo.Users t',
```
Comment (NicholasYip) [inline backend/controllers/userController.js]: we should agree on using either only sequelize, or only writing the queries ourselves, having a mix of both makes things a bit messy

File: frontend/components/admin/UserLists.tsx [hunk 9]
```diff
@@ -0,0 +1,76 @@
+import { Box, useDisclosure, Spinner, SimpleGrid } from '@chakra-ui/react';
+import { Fragment, useState } from 'react';
+import UserList from './UserList';
+import UserInfoSimple from '@frontend/models/UserInfoSimple';
+import UserModal from './UserModal';
+import useRole from '@frontend/hooks/userRole';
+
+const UserLists = () => {
+  const { userRoles, isLoading, isError } = useRole(); // Custom Hook to Fetch the user/role data
```
Comment (NicholasYip) [inline frontend/components/admin/UserLists.tsx:13]: good stuff with the custom hooks

**Review by NicholasYip — APPROVED** [review]
great work
left some small recommendations

File: backend/routes/admins.js [hunk 10]
```diff
@@ -1,27 +1,33 @@
...
+  (req, res) => {
+    res.send('If you can access this then you are an admin !');
+  }
+);
 
 //Update user roles as admin
-router.put('/updateRole', authUser.verifyUser, authUser.verifyRole(constants.ROLE.ADMIN), (req, res) => {
-    adminController.updateRole(req,res);
-})
+//TODO Fix route, 403 status code
```
Comment (jfv-2000) [inline backend/routes/admins.js]: I found the 403 status code, it's associated with the middleware, due to the naming convention difference. Fixing the req,body naming convention from req.body.AccountId to req.body accountId should fix this.

File: backend/routes/users.js [hunk 11]
```diff
@@ -11,6 +13,9 @@ router.get('/getUser', authJwt.verifyToken, (req, res) => {
   authController.getAuthUser(req, res);
 });
 
+router.get('/role', userController.getAllRoles);
+router.patch('/role', adminController.updateRole);
```
Comment (jfv-2000) [inline backend/routes/users.js]: The route updateRole (line 17) is already called in the admin route, can we remove it if not necessary.

**Review by jfv-2000 — CHANGES_REQUESTED** [review]
Good work ! There are some important functionality bugs that should be fixed in the back-end

File: backend/package-lock.json [hunk 12]
```diff
@@ -1,7 +1,7 @@
 {
```
Comment (amqkv) [inline backend/package-lock.json:1]: No need to track package-lock to the repo so we should remove it

File: frontend/components/admin/UserLists.tsx [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+
+  const { isOpen, onOpen, onClose } = useDisclosure(); // Hook to deal with the modal visibility
+
+  const [userSelected, setUserSelected] = useState<UserInfoSimple>({
+    AccountId: 0,
+    FirstName: '',
+    LastName: '',
+    Role: '',
+  });
```
Comment (jfv-2000) [inline frontend/components/admin/UserLists.tsx:22]: Match the correct naming convention as mentionned in other comments for the back-end files, refer to the signUpParameters interface in the minicap/frontend/functions/signUp.ts file

**Review by jfv-2000 — CHANGES_REQUESTED** [review]

File: frontend/pages/admin/user-list.tsx [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+    props: {
+      session: await getSession(context),
+    },
+  };
+}
+
+const UserListPage = () => {
+  const { data: session } = useSession();
+
+  if (session?.user.Role === 'Admin') return <UserLists />;
```
Comment (jfv-2000) [inline frontend/pages/admin/user-list.tsx]: If possible, have a folder of constants instead of using hardcoded values

**Review by jfv-2000 — CHANGES_REQUESTED** [review]

File: backend/package.json (hunk 6 above)
Comment (SergioAAI) [reply backend/package.json]: On my side , node crashes without tedious

File: backend/routes/admins.js (hunk 10 above)
Comment (SergioAAI) [reply backend/routes/admins.js]: Turns into 401 with the naming convention changes

File: backend/routes/users.js (hunk 11 above)
Comment (SergioAAI) [reply backend/routes/users.js]: The Admin route doesn't properly work so this route is used for the moment

File: backend/routes/admins.js (hunk 10 above)
Comment (jfv-2000) [reply backend/routes/admins.js]: It's the error of one of the comments I mentionned above in the file adminController.js: "Tested the update function locally, currently not working properly due to line 16. The accountId of the user should be used to verify if the user has the rights to update roles as an admin. The email should be used in this case to properly update the role of a user."

**Pull request merged**